pgjdbc/src/main/java/org/postgresql/core/v2/ProtocolConnectionImpl.java

```diff
@@ -197,6 +209,15 @@ synchronized void addWarning(SQLWarning newWarning) {
 
   synchronized void addNotification(PGNotification notification) {
     notifications.add(notification);
```

Comment:
I'd be inclined to make it an either/or proposition.  If you establish a listener we don't queue up the notifications - only raise them via event.  If no listeners are present we continue with the present logic.